airbyte-ci/connectors/regression-testing/src/regression_testing/connector_runner.py

```diff
@@ -0,0 +1,378 @@
+
+        Returns:
+            dagger.Container: The dagger container with the environment variables set.
+        """
+        for k, v in env_vars.items():
+            container = container.with_env_variable(k, str(v))
+        return container
+
+    async def call_spec(self, raise_container_error=False) -> List[AirbyteMessage]:
+        return await self._run(["spec"], raise_container_error)
+
+    async def call_check(self, config: SecretDict, raise_container_error: bool = False) -> List[AirbyteMessage]:
+        return await self._run(
+            ["check", "--config", self.IN_CONTAINER_CONFIG_PATH],
+            raise_container_error,
+            config=config,
+        )
+
+    async def call_discover(self, config: SecretDict, raise_container_error: bool = False) -> List[AirbyteMessage]:
+        return await self._run(
+            ["discover", "--config", self.IN_CONTAINER_CONFIG_PATH],
+            raise_container_error,
+            config=config,
+        )
+
+    async def call_read(
+        self, config: SecretDict, catalog: ConfiguredAirbyteCatalog, raise_container_error: bool = False, enable_caching: bool = True
+    ) -> List[AirbyteMessage]:
+        return await self._run(
+            ["read", "--config", self.IN_CONTAINER_CONFIG_PATH, "--catalog", self.IN_CONTAINER_CATALOG_PATH],
+            raise_container_error,
+            config=config,
+            catalog=catalog,
+            enable_caching=enable_caching,
+        )
+
+    async def call_read_with_state(
+        self,
+        config: SecretDict,
+        catalog: ConfiguredAirbyteCatalog,
+        state: dict,
+        raise_container_error: bool = False,
+        enable_caching: bool = True,
+    ) -> List[AirbyteMessage]:
+        return await self._run(
+            [
+                "read",
+                "--config",
+                self.IN_CONTAINER_CONFIG_PATH,
+                "--catalog",
+                self.IN_CONTAINER_CATALOG_PATH,
+                "--state",
+                self.IN_CONTAINER_STATE_PATH,
+            ],
+            raise_container_error,
+            config=config,
+            catalog=catalog,
+            state=state,
+            enable_caching=enable_caching,
+        )
+
+    async def get_container_env_variable_value(self, name: str) -> str:
+        return await self._connector_under_test_container.env_variable(name)
+
+    async def get_container_label(self, label: str):
+        return await self._connector_under_test_container.label(label)
+
+    async def get_container_entrypoint(self):
+        entrypoint = await self._connector_under_test_container.entrypoint()
+        return " ".join(entrypoint)
+
+    async def _run(
+        self,
+        airbyte_command: List[str],
+        raise_container_error: bool,
+        config: SecretDict = None,
+        catalog: dict = None,
+        state: Union[dict, list] = None,
+        enable_caching=True,
+    ) -> List[AirbyteMessage]:
```

Comment:
It's not returning a list of AirbyteMessage if I'm not mistaken 😄

Reply:
What do you think about making `_run` return a tuple with the executed command and the awaited container.

Then this tuple can be passed to a "backend" which would produce the artifact we would need downstream for debugging or testing?

This would help making this class self contained:  only running airbyte commands on connector containers